pycheribuild/config/compilation_targets.py

```diff
@@ -210,18 +215,42 @@ def riscv_arch_string(self):
     @property
     def riscv_abi(self):
         assert self.target.is_riscv(include_purecap=True)
+
+        if self.is_baremetal():
+            return self.riscv_softfloat_abi()  # Baremetal/FreeRTOS only supports softfloat
+
         if self.target.is_cheri_purecap():
-            return "l64pc128d"  # 64-bit double-precision hard-float + purecap
+            if self.target.cpu_architecture == CPUArchitecture.RISCV32:
+                # 32-bit double-precision hard-float + purecap
+                return "il32pc64d"
+            else:
+                # 64-bit double-precision hard-float + purecap
+                return "l64pc128d"
         else:
-            return "lp64d"  # 64-bit double-precision hard-float
+            if self.target.cpu_architecture == CPUArchitecture.RISCV32:
+                # 32-bit double-precision hard-float
+                return "ilp32d"
+            else:
+                # 64-bit double-precision hard-float
+                return "lp64d"
 
     @property
     def riscv_softfloat_abi(self):
         assert self.target.is_riscv(include_purecap=True)
         if self.target.is_cheri_purecap():
-            return "l64pc128"  # 64-bit soft-float purecap
+            if self.target.cpu_architecture == CPUArchitecture.RISCV32:
```

Comment:
Purecap and hybrid/vanilla are now round the wrong way, and the comments have got worse.